src/sources/exec.rs

```diff
@@ -0,0 +1,843 @@
+        EXEC
+    }
+}
+
+pub fn run_scheduled(
+    config: ExecConfig,
+    exec_interval_secs: u64,
+    shutdown: ShutdownSignal,
+    out: Pipeline,
+) -> crate::Result<super::Source> {
+    Ok(Box::pin(async move {
+        info!("Starting scheduled exec runs.");
+        let schedule = time::Duration::from_secs(exec_interval_secs);
+        let mut interval = time::interval(schedule).take_until(shutdown.clone());
+
+        while interval.next().await.is_some() {
+            // Mark the start time just before spawning the process as
+            // this seems to be the best approximation of exec duration
+            let now = Instant::now();
+
+            // Wait for our task to finish, wrapping it in a timeout
+            let timeout = tokio::time::timeout(
+                schedule,
+                run_command(config.clone(), shutdown.clone(), out.clone()),
+            );
+
+            let timeout_result = timeout.await;
+
+            match timeout_result {
+                Ok(output) => {
+                    if let Err(command_error) = output {
+                        emit!(ExecFailed {
+                            command: config.command.as_str(),
+                            error: command_error,
+                        });
+                    }
+                }
+                Err(_) => {
+                    emit!(ExecTimeout {
+                        command: config.command.as_str(),
+                        elapsed_seconds: now.elapsed().as_secs(),
+                    });
+                }
+            }
+        }
+
+        info!("Finished scheduled exec runs.");
+        Ok(())
+    }))
+}
+
+pub fn run_streaming(
+    config: ExecConfig,
+    respawn_on_exit: bool,
+    respawn_interval_secs: u64,
+    shutdown: ShutdownSignal,
+    out: Pipeline,
+) -> crate::Result<super::Source> {
+    Ok(Box::pin(async move {
+        if respawn_on_exit {
+            let duration = time::Duration::from_secs(respawn_interval_secs);
+
+            // Continue to loop while shutdown is pending
+            while futures::poll!(shutdown.clone()).is_pending() {
+                let output = run_command(config.clone(), shutdown.clone(), out.clone()).await;
+
+                if let Err(command_error) = output {
+                    emit!(ExecFailed {
+                        command: config.command.as_str(),
+                        error: command_error,
+                    });
+                }
+
+                if futures::poll!(shutdown.clone()).is_ready() {
+                    break;
+                } else {
+                    warn!("Streaming processed ended before shutdown.");
+
+                    // Using time interval so it can utilize the shutdown signal
+                    let mut interval = time::interval(duration).take_until(shutdown.clone());
+
+                    // Call next twice since the first is immediate
+                    interval.next().await;
+                    interval.next().await;
```

Comment:
Please elaborate here. Why is the first immediate? Could this use `interval_at` instead to eliminate the need for the first wait?

Reply:
I have changed this logic to use `tokio::select!` and `delay_for`. I think the first was immediate because of the underlying [interval](https://docs.rs/tokio/0.2.25/tokio/time/fn.interval.html) implementation, I mentions:-

> Creates new Interval that yields with interval of duration. The first tick completes immediately.